File: MarioKart/MarioKart/TWGL/TWGLNode.m
//
//  RWTModel.m
//  HelloOpenGL
//
//  Created by Main Account on 3/18/14.
//  Copyright (c) 2014 Razeware LLC. All rights reserved.
//

#import "TWGLNode.h"
#import "TWGLShadersReference.h"
#import <OpenGLES/ES2/glext.h>

@implementation TWGLNode {
    char *_name;
    GLuint _vao;
    GLuint _vertexBuffer;
    unsigned int _vertexCount;
    TWGLShadersReference *_shader;
}

- (instancetype)initWithName:(char *)name shader:(TWGLShadersReference *)shader vertices:(GLfloat *)vertices vertexCount:(unsigned int)vertexCount {
    
    if (self = [super init]) {
        
        _name = name;
        _vertexCount = vertexCount;
        _shader = shader;
        self.position = GLKVector3Make(0, 0, 0);
        self.rotationX = 0;
        self.rotationY = 0;
        self.rotationZ = 0;
        self.scale = 1.0;
        self.children = [NSMutableArray array];
        
        glGenVertexArraysOES(1, &_vao);
        glBindVertexArrayOES(_vao);
        
        // Generate vertex buffer
        glGenBuffers(1, &_vertexBuffer);
        glBindBuffer(GL_ARRAY_BUFFER, _vertexBuffer);
        glBufferData(GL_ARRAY_BUFFER, vertexCount * 3 * sizeof(GL_FLOAT), vertices, GL_STATIC_DRAW);
//        glBufferData(GL_ARRAY_BUFFER, vertexCount * sizeof(GL_FLOAT), vertices, GL_STATIC_DRAW);
        
//        // Generate index buffer
//        glGenBuffers(1, &_indexBuffer);
//        glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, _indexBuffer);
//        glBufferData(GL_ELEMENT_ARRAY_BUFFER, indexCount * sizeof(GLubyte), indices, GL_STATIC_DRAW);
        
        // Enable vertex attributes
        glEnableVertexAttribArray(TWGLVertexAttribPosition);
//        glVertexAttribPointer(TWGLVertexAttribPosition, 3, GL_FLOAT, GL_FALSE, sizeof(GL_FLOAT), (const GLvoid *) offsetof(TWGLVertex, Position));
        glVertexAttribPointer(TWGLVertexAttribPosition, 3, GL_FLOAT, GL_FALSE, 3*sizeof(GL_FLOAT), 0);
//        glVertexAttribPointer(TWGLVertexAttribPosition, 3, GL_FLOAT, GL_FALSE, sizeof(GL_FLOAT), 0);
//        glEnableVertexAttribArray(TWGLVertexAttribColor);
//        glVertexAttribPointer(TWGLVertexAttribColor, 4, GL_FLOAT, GL_FALSE, sizeof(TWGLVertex), (const GLvoid *) offsetof(TWGLVertex, Color));
//        glEnableVertexAttribArray(TWGLVertexAttribTexCoord);
//        glVertexAttribPointer(TWGLVertexAttribTexCoord, 2, GL_FLOAT, GL_FALSE, sizeof(TWGLVertex), (const GLvoid *) offsetof(TWGLVertex, TexCoord));
//        glEnableVertexAttribArray(TWGLVertexAttribNormal);
//        glVertexAttribPointer(TWGLVertexAttribNormal, 3, GL_FLOAT, GL_FALSE, sizeof(TWGLVertex), (const GLvoid *) offsetof(TWGLVertex, Normal));
        
        glBindVertexArrayOES(0);
        glBindBuffer(GL_ARRAY_BUFFER, 0);
        glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, 0);
        
    }
    return self;
}

- (GLKMatrix4)modelMatrix {
    GLKMatrix4 modelMatrix = GLKMatrix4Identity;
    modelMatrix = GLKMatrix4Translate(modelMatrix, self.position.x, self.position.y, self.position.z);
    modelMatrix = GLKMatrix4Rotate(modelMatrix, self.rotationX, 1, 0, 0);
    modelMatrix = GLKMatrix4Rotate(modelMatrix, self.rotationY, 0, 1, 0);
    modelMatrix = GLKMatrix4Rotate(modelMatrix, self.rotationZ, 0, 0, 1);
    modelMatrix = GLKMatrix4Scale(modelMatrix, self.scale, self.scale, self.scale);
    return modelMatrix;
}

- (void)renderWithParentModelViewMatrix:(GLKMatrix4)parentModelViewMatrix {
    
    GLKMatrix4 modelViewMatrix = GLKMatrix4Multiply(parentModelViewMatrix, [self modelMatrix]);
    for (TWGLNode *child in self.children) {
        [child renderWithParentModelViewMatrix:modelViewMatrix];
    }
    
    _shader.modelViewMatrix = modelViewMatrix;
    _shader.texture = self.texture;
    [_shader prepareToDraw];
    
    glBindVertexArrayOES(_vao);
    glDrawArrays(GL_TRIANGLES, 0, _vertexCount);
    glBindVertexArrayOES(0);
    
}

- (void)updateWithDelta:(NSTimeInterval)dt {
    for (TWGLNode *child in self.children) {
        [child updateWithDelta:dt];
    }
}

- (void)loadTexture:(NSString *)filename {
    NSError *error;
    NSString *path = [[NSBundle mainBundle] pathForResource:filename ofType:nil];
    
    NSDictionary *options = @{ GLKTextureLoaderOriginBottomLeft: @YES };
    GLKTextureInfo *info = [GLKTextureLoader textureWithContentsOfFile:path options:options error:&error];
    if (info == nil) {
        NSLog(@"Error loading file: %@", error.localizedDescription);
    } else {
        self.texture = info.name;
    }
}

@end

File: MarioKart/MarioKart/TWGL/TWGLNode.swift
//
//  TWGLNode.swift
//  MarioKart
//
//  Created by Txai Wieser on 11/3/15.
//  Copyright © 2015 Txai Wieser. All rights reserved.
//

import GLKit
import OpenGLES

class TWGLNode {
    let name:String
    var vao = GLuint()
    var vertexBuffer = GLuint()
    let vertexCount:Int
    weak var shader:TWGLShaderReference?
    
    var texture = GLuint()
    //    var size:GLKVector3 = 0
    
    
    var position:GLKVector3 = GLKVector3Make(0, 0, 0)
    var rotation:GLKVector3 = GLKVector3Make(0, 0, 0)
    var scale:GLfloat = 1.0
    var matColor:GLKVector4 = GLKVector4Make(1, 1, 1, 1)
    var children:[TWGLNode] = []
    
    
    func loadTexture(filename:String) {
        let path = NSBundle.mainBundle().pathForResource(filename, ofType: nil)
        
        
        let options = [GLKTextureLoaderOriginBottomLeft : true]
        
        do {
            let info = try GLKTextureLoader.textureWithContentsOfFile(path!, options: options)
            self.texture = info.name
        } catch let error as NSError {
            print("Error loading file: %@", error.localizedDescription)
        }
    }
    
    
    init(name:String, texture:String?, shader:TWGLShaderReference, vertices:[TWGLVertexInfo]) {
        self.name = name
        self.shader = shader
        self.vertexCount = vertices.count
        
        glGenVertexArraysOES(1, &vao)
        glBindVertexArrayOES(vao)
        
        // Generate vertex buffer
        glGenBuffers(1, &vertexBuffer)
        glBindBuffer(GLenum(GL_ARRAY_BUFFER), vertexBuffer)
        glBufferData(GLenum(GL_ARRAY_BUFFER), vertexCount * sizeof(TWGLVertexInfo), vertices, GLenum(GL_STATIC_DRAW))
        
        // Enable vertex attributes
        glEnableVertexAttribArray(GLuint(GLKVertexAttrib.Position.rawValue))
        glVertexAttribPointer(GLuint(GLKVertexAttrib.Position.rawValue), 3, GLenum(GL_FLOAT), GLboolean(GL_FALSE),  GLsizei(sizeof(TWGLVertexInfo)), BUFFER_OFFSET(0))
        glEnableVertexAttribArray(GLuint(GLKVertexAttrib.Color.rawValue))
        glVertexAttribPointer(GLuint(GLKVertexAttrib.Color.rawValue), 4, GLenum(GL_FLOAT), GLboolean(GL_FALSE), GLsizei(sizeof(TWGLVertexInfo)), BUFFER_OFFSET(3))
        glEnableVertexAttribArray(GLuint(GLKVertexAttrib.TexCoord0.rawValue))
        glVertexAttribPointer(GLuint(GLKVertexAttrib.TexCoord0.rawValue), 2, GLenum(GL_FLOAT), GLboolean(GL_FALSE), GLsizei(sizeof(TWGLVertexInfo)), BUFFER_OFFSET((3+4)))
        glEnableVertexAttribArray(GLuint(GLKVertexAttrib.Normal.rawValue))
        glVertexAttribPointer(GLuint(GLKVertexAttrib.Normal.rawValue), 3, GLenum(GL_FLOAT), GLboolean(GL_FALSE), GLsizei(sizeof(TWGLVertexInfo)), BUFFER_OFFSET((3+4+3)))
        
        glBindVertexArrayOES(0)
        glBindBuffer(GLenum(GL_ARRAY_BUFFER), 0)
        glBindBuffer(GLenum(GL_ELEMENT_ARRAY_BUFFER), 0)
        
        if let t = texture {
            loadTexture(t)
        }
    }
    
    func modelMatrix() -> GLKMatrix4 {
        var modelMatrix = GLKMatrix4Identity
        modelMatrix = GLKMatrix4Translate(modelMatrix, position.x, position.y, position.z)
        modelMatrix = GLKMatrix4Rotate(modelMatrix, rotation.x, 1, 0, 0)
        modelMatrix = GLKMatrix4Rotate(modelMatrix, rotation.y, 0, 1, 0)
        modelMatrix = GLKMatrix4Rotate(modelMatrix, rotation.z, 0, 0, 1)
        modelMatrix = GLKMatrix4Scale(modelMatrix, scale, scale, scale)
        return modelMatrix
    }
    
    func render(parentModelViewMatrix:GLKMatrix4) {
        let modelViewMatrix = GLKMatrix4Multiply(parentModelViewMatrix, modelMatrix())
        children.forEach { $0.render(modelViewMatrix) }
        
        shader?.modelViewMatrix = modelViewMatrix
        shader?.texture = self.texture
        shader?.matColor = self.matColor
        shader?.prepareToDraw()
        
        glBindVertexArrayOES(vao)
        glDrawArrays(GLenum(GL_TRIANGLES), 0, Int32(vertexCount))
        glBindVertexArrayOES(0)
    }
    
    func update(delta:NSTimeInterval) {
        children.forEach { $0.update(delta) }
    }
    //
    //
    //    - (void)touchesBegan:(NSSet *)touches withEvent:(UIEvent *)event {
    //    }
    //    - (void)touchesMoved:(NSSet *)touches withEvent:(UIEvent *)event {
    //    }
    //    - (void)touchesEnded:(NSSet *)touches withEvent:(UIEvent *)event {
    //    }
    
    //    - (CGRect)boundingBoxWithModelViewMatrix:(GLKMatrix4)parentModelViewMatrix {
    //
    //    GLKMatrix4 modelViewMatrix = GLKMatrix4Multiply(parentModelViewMatrix, [self modelMatrix])
    //
    //    GLKVector4 lowerLeft = GLKVector4Make(-self.width/2, -self.height/2, 0, 1)
    //    lowerLeft = GLKMatrix4MultiplyVector4(modelViewMatrix, lowerLeft)
    //    GLKVector4 upperRight = GLKVector4Make(self.width/2, self.height/2, 0, 1)
    //    upperRight = GLKMatrix4MultiplyVector4(modelViewMatrix, upperRight)
    //
    //    CGRect boundingBox = CGRectMake(lowerLeft.x, lowerLeft.y, upperRight.x - lowerLeft.x, upperRight.y - lowerLeft.y)
    //    return boundingBox
    
}